src/bases/DataProviderBase.ts

```diff
@@ -0,0 +1,86 @@
+import { Observable, Observer } from '@dojo/core/Observable';
+import { DataProperties, SortDetails } from '../interfaces';
+
+export interface DataProviderConfiguration {
+	sort?: SortDetails | SortDetails[];
+}
+
+export interface Options {
+	[option: string]: any;
+	configuration?: DataProviderConfiguration;
+}
+
+export interface DataProviderState<O extends Options> {
+	options: O;
+	sort?: SortDetails[];
+}
+
+abstract class DataProviderBase<T, O extends Options> {
+	private data: DataProperties<T>;
+	private state: DataProviderState<O>;
+
+	observable: Observable<DataProperties<T>>;
+	observers: Observer<DataProperties<T>>[];
+
+	constructor(options: O) {
+		const {
+			configuration: {
+				sort = []
+			} = {}
+		} = options;
+
+		this.state = {
+			options
+		};
+		if (sort) {
+			this.state.sort = Array.isArray(sort) ? sort : [ sort ];
+		}
+		this.observers = [];
+		this.observable = new Observable((observer: Observer<DataProperties<T>>) => {
+			this.observers.push(observer);
+			if (this.data) {
+				observer.next(this.data);
+			}
+		});
+	}
+
+	protected abstract buildData(state: DataProviderState<O>): DataProperties<T>;
+
+	configure({ sort }: DataProviderConfiguration) {
```

Comment:
with public methods, I like to see the return type being explicit instead of being inferred, I don't care about private functions/methods, but public ones, you are essentially being explicit about your external API and the explicit return types helps make sure your code block does return the right type (and you don't have to load it into an editor to see what the return type is)